### PR TITLE
[DPE-3267] - feat: in-place upgrades

### PR DESCRIPTION
**Pull request by marcoppenheimer**
#### `feat: add in-place upgrade support`
- Disabled the `resume-upgrade` functionality from the lib, to keep behavior the same between VM + K8s
- Although it works, upgrades on K8s needs to be manually tested more thoroughly before merging. NOTE - int test is skipped for now, as we don't have a published `stable` charm to test against
- No unit-tests right now, as they're currently being refactored

### CODE REVIEW

File: src/events/upgrade.py [hunk 1]
```diff
@@ -33,85 +38,108 @@ class KafkaDependencyModel(BaseModel):
...
-        units = set([self.charm.unit] + list(self.charm.peer_relation.units))  # type: ignore[reportOptionalMemberAccess]
-        for unit in units:
-            upgrade_stack.append(int(unit.name.split("/")[-1]))
-
-        return upgrade_stack
-
-    @override
-    def log_rollback_instructions(self) -> None:
-        logger.critical(ROLLBACK_INSTRUCTIONS)
+    def _on_kafka_pebble_ready_upgrade(self, event: EventBase) -> None:
```
Comment (zmraul) [inline src/events/upgrade.py:53]: **question:** Why can't this hook follow the same pattern as vm? I mainly mean the call to `self.charm._on_update_status(event)` If that call is needed then my other comment about update_status check can be ignored.

File: src/events/upgrade.py [hunk 2]
```diff
@@ -33,85 +38,108 @@ class KafkaDependencyModel(BaseModel):
...
 
-        logger.info(f"{self.charm.unit.name} upgrading service...")
-        self.charm.workload.restart()
+        self.set_unit_completed()
 
-        try:
-            logger.debug("Running post-upgrade check...")
-            self.post_upgrade_check()
+    @property
+    def idle(self) -> bool:
```
Comment (zmraul) [inline src/events/upgrade.py:97]: This is different than vm. Should we change it back there?

File: src/events/upgrade.py [hunk 3]
```diff
@@ -33,85 +38,108 @@ class KafkaDependencyModel(BaseModel):
...
     def __init__(self, charm: "KafkaK8sCharm", **kwargs):
         super().__init__(charm, **kwargs)
         self.charm = charm
+        self.substrate = self.charm.state.substrate
 
-    @property
-    def current_version(self) -> str:
-        """Get current Kafka version."""
-        dependency_model: DependencyModel = getattr(self.dependency_model, "kafka_service")
-        return dependency_model.version
```
Comment (zmraul) [inline src/events/upgrade.py:47]: **todo:** This property needs to be used. The upgrade is missing the `inter.broker.protocol.version` addition to `server.properties`. [See here](https://github.com/canonical/kafka-operator/pull/127/files#diff-24a3fe5c783755412cb9de83238b4dfd0699345bf50dcadc3ec23c6749361592R456-R465)

**Review by zmraul — CHANGES_REQUESTED** [review]
Nice! Sorry for blocking with the refactor here. I left some comments. There is one piece that is missing from the PR, see comment about `inter.broker.protocol.version`

**Comment by marcoppenheimer** [comment]
> Sorry for blocking with the refactor here

No worries! Was just getting it out early for review, helpful comments :). I'll address these next week.

**Review by zmraul — CHANGES_REQUESTED** [review]
Related to the `inter.broker.protocol.version` field:
```
$ diff -C 10 src/managers/config.py ../kafka-operator/src/managers/config.py 
*** src/managers/config.py      2024-01-29 17:36:01.183034389 +0100
--- ../kafka-operator/src/managers/config.py    2024-01-29 17:43:14.819713568 +0100
***************
*** 97,120 ****
--- 97,122 ----
  
  
  class KafkaConfigManager:
      """Manager for handling Kafka configuration."""
  
      def __init__(
          self,
          state: ClusterState,
          workload: WorkloadBase,
          config: CharmConfig,
+         current_version: str,
      ):
          self.state = state
          self.workload = workload
          self.config = config
+         self.current_version = current_version
  
      @property
      def log_level(self) -> str:
          """Return the Java-compliant logging level set by the user.
  
          Returns:
              String with these possible values: DEBUG, INFO, WARN, ERROR
          """
          # Remapping to WARN that is generally used in Java applications based on log4j and logback.
          if self.config.log_level == LogLevel.WARNING.value:
***************
*** 305,324 ****
--- 307,337 ----
              Listener(host=self.state.broker.host, protocol=auth, scope="CLIENT")
              for auth in self.auth_mechanisms
          ]
  
      @property
      def all_listeners(self) -> list[Listener]:
          """Return a list with all expected listeners."""
          return [self.internal_listener] + self.client_listeners
  
      @property
+     def inter_broker_protocol_version(self) -> str:
+         """Creates the protocol version from the kafka version.
+ 
+         Returns:
+             String with the `major.minor` version
+         """
+         # Remove patch number from full vervion.
+         major_minor = self.current_version.split(".", maxsplit=2)
+         return ".".join(major_minor[:2])
+ 
+     @property
      def rack_properties(self) -> list[str]:
          """Builds all properties related to rack awareness configuration.
  
          Returns:
              List of properties to be set
          """
          # TODO: not sure if we should make this an instance attribute like the other paths
          rack_path = f"{self.workload.paths.conf_path}/rack.properties"
          return self.workload.read(rack_path) or []
  
***************
*** 364,383 ****
--- 377,397 ----
          advertised_listeners = [listener.advertised_listener for listener in self.all_listeners]
  
          properties = (
              [
                  f"super.users={self.state.super_users}",
                  f"log.dirs={self.state.log_dirs}",
                  f"listener.security.protocol.map={','.join(protocol_map)}",
                  f"listeners={','.join(listeners_repr)}",
                  f"advertised.listeners={','.join(advertised_listeners)}",
                  f"inter.broker.listener.name={self.internal_listener.name}",
+                 f"inter.broker.protocol.version={self.inter_broker_protocol_version}",
              ]
              + self.config_properties
              + self.scram_properties
              + self.default_replication_properties
              + self.auth_properties
              + self.rack_properties
              + DEFAULT_CONFIG_OPTIONS.split("\n")
          )
  
          if self.state.cluster.tls_enabled and self.state.broker.certificate:
```

**Review by zmraul — APPROVED** [review]

**Pull request merged**